opentreemap/treemap/udf.py

```diff
@@ -358,17 +369,24 @@ def _validate_and_update_choice(
                 {'datatype': [_("Can't change choices "
                                 "on a non-choice field")]})
 
-        if old_choice_value not in datatype['choices']:
+        choices = datatype['choices']
+        protected_choices = datatype.get('protected_choices', [])
+        all_choices = choices + protected_choices
+
+        if old_choice_value not in all_choices:
             raise ValidationError(
                 {'datatype': [_("Choice '%(choice)s' not found") % {
                     'choice': old_choice_value}]})
 
-        choices = datatype['choices']
-        if new_choice_value:
-            choices = [c if c != old_choice_value else new_choice_value
-                       for c in choices]
-        else:
-            choices = [c for c in choices if c != old_choice_value]
+        if old_choice_value in protected_choices:
+            raise ValidationError({
+                'datatype': [
+                    _("Can't modify protected choice '%(choice)s'") % {
+                        'choice': old_choice_value}]})
+
+        choices = self._list_replace_or_remove(choices,
+                                               old_choice_value,
+                                               new_choice_value)
```

Comment:
This change isn't related to the protected UDFs work. I just noticed that this logic was already encapsulated in an existing helper function and thought it would be a good idea to use it here too.